crates/oxc_parser/src/js/expression.rs

```diff
@@ -1427,40 +1433,44 @@ impl<'a> ParserImpl<'a> {
 
     fn is_await_expression(&mut self) -> bool {
         if self.at(Kind::Await) {
-            let peek_token = self.peek_token();
-            // Allow arrow expression `await => {}`
-            if peek_token.kind() == Kind::Arrow {
-                return false;
-            }
             if self.ctx.has_await() {
                 return true;
             }
-            // The following expressions are ambiguous
-            // await + 0, await - 0, await ( 0 ), await [ 0 ], await / 0 /u, await ``, await of []
-            if matches!(
-                peek_token.kind(),
-                Kind::Of | Kind::LParen | Kind::LBrack | Kind::Slash | Kind::RegExp
-            ) {
-                return false;
-            }
-
-            return !peek_token.is_on_new_line() && peek_token.kind().is_after_await_or_yield();
+            return self.lookahead(|p| {
+                Self::next_token_is_identifier_or_keyword_or_literal_on_same_line(p, true)
+            });
         }
         false
     }
 
     fn is_yield_expression(&mut self) -> bool {
         if self.at(Kind::Yield) {
-            let peek_token = self.peek_token();
-            // Allow arrow expression `yield => {}`
-            if peek_token.kind() == Kind::Arrow {
-                return false;
-            }
             if self.ctx.has_yield() {
                 return true;
             }
-            return !peek_token.is_on_new_line() && peek_token.kind().is_after_await_or_yield();
+            return self.lookahead(|p| {
+                Self::next_token_is_identifier_or_keyword_or_literal_on_same_line(p, false)
+            });
         }
         false
     }
+
+    fn next_token_is_identifier_or_keyword_or_literal_on_same_line(
+        &mut self,
+        is_await: bool,
+    ) -> bool {
+        self.bump_any();
+
+        // NOTE: This check implementation is based on TypeScript parser.
+        // But TS does not have this exception.
+        // This is needed to pass parser_babel test to parse:
+        // `for (await of [])` with `sourceType: script`
+        if !self.is_ts && is_await && self.at(Kind::Of) {
+            return false;
+        }
```

Comment:
```rs
if matches!(
    peek_token.kind(),
    Kind::Of | Kind::LParen | Kind::LBrack | Kind::Slash | Kind::RegExp
) {
```

The original code covered these.
But even with original code, the test did not fail when I erased all but the `Of`.